flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -188,6 +221,77 @@ public void checkpointState(
 		}
 	}
 
+	private OperatorSnapshotFutures buildOperatorSnapshotFutures(
+			CheckpointMetaData checkpointMetaData,
+			CheckpointOptions checkpointOptions,
+			OperatorChain<?, ?> operatorChain,
+			StreamOperator<?> op,
+			Supplier<Boolean> isCanceled,
+			ChannelStateWriteResult channelStateWriteResult) throws Exception {
+		CheckpointStreamFactory storage = checkpointStorage.resolveCheckpointStorageLocation(
+			checkpointMetaData.getCheckpointId(),
+			checkpointOptions.getTargetLocation());
+		OperatorSnapshotFutures snapshotInProgress = checkpointStreamOperator(
+			op,
+			checkpointMetaData,
+			checkpointOptions,
+			storage,
+			isCanceled);
+		if (op == operatorChain.getHeadOperator()) {
+			snapshotInProgress.setInputChannelStateFuture(
+				channelStateWriteResult
+					.getInputChannelStateHandles()
+					.thenApply(StateObjectCollection::new)
+					.thenApply(SnapshotResult::of));
+		}
+		if (op == operatorChain.getTailOperator()) {
+			snapshotInProgress.setResultSubpartitionStateFuture(
+				channelStateWriteResult
+					.getResultSubpartitionStateHandles()
+					.thenApply(StateObjectCollection::new)
+					.thenApply(SnapshotResult::of));
+		}
+		return snapshotInProgress;
+	}
+
+	@Override
+	public void close() throws IOException {
+		channelStateWriter.close();
+	}
+
+	// Caches checkpoint output stream factories to prevent multiple output stream per checkpoint.
+	// This could result from requesting output stream by different entities (this and channelStateWriter)
+	// We can't just pass a stream to the channelStateWriter because it can receive checkpoint call earlier than this class
+	// in some unaligned checkpoints scenarios
+	private static class CachingCheckpointStorageWorkerView implements CheckpointStorageWorkerView {
```

Comment:
Do we have the tests coverage for this new introduced caching function in the commit "[FLINK-16744][task] send channel state handles to JM"?